### PR TITLE
Rates matrix solver

### PR DESCRIPTION
**Pull request by andrewfullard**
### :pencil: Description

**Type:** :rocket: `feature` 

Initial plans for replacing the old "NLTE" setup with a hopefully cleaner structure.

### :vertical_traffic_light: Testing

How did you test these changes?

- [ ] Testing pipeline
- [ ] Other method (describe)
- [ ] My changes can't be tested (explain why)

### :ballot_box_with_check: Checklist

- [ ] I requested two reviewers for this pull request
- [ ] I updated the documentation according to my changes
- [ ] I built the documentation by applying the `build_docs` label

> **Note:** If you are not allowed to perform any of these actions, ping (@) a contributor.

### CODE REVIEW

File: tardis/plasma/electron_distribution/base.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+from astropy import units as u
+
+
+@dataclass
+class ElectronDistribution:
+    """Electron temperature and density distribution.
+
+    temperature : Quantity
+        Electron temperatures.
+    number_density : Quantity
```
Comment (jvshields) [inline tardis/plasma/electron_distribution/base.py:12]: Being that we often just take .value of these quantities, is it worth specifying that these should probably be in cgs? Or is that taken care of elsewhere?

Comment (andrewfullard) [reply tardis/plasma/electron_distribution/base.py:12]: You're correct, I think CGS specification is reasonable

**Comment by wkerzendorf** [comment]
adding more tests to get coverage up probably

File: tardis/plasma/electron_distribution/base.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+from dataclasses import dataclass
+
+from astropy import units as u
+
+
+@dataclass
+class ElectronDistribution:
```
Comment (wkerzendorf) [inline tardis/plasma/electron_distribution/base.py:7]: `class ElectronEnergyDistribution` 

`class MaxwellElectronEnergyDistribution` as a subclass

Comment (wkerzendorf) [reply tardis/plasma/electron_distribution/base.py:7]: also `ElectronEnergySpectrum`

File: tardis/plasma/equilibrium/rate_matrix.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        levels : pd.DataFrame
+            DataFrame of energy levels.
+        """
+        self.rate_solvers = rate_solvers
+        self.levels = levels
+
+    def solve(
+        self,
+        radiation_field,
+        electron_distribution,
```
Comment (wkerzendorf) [inline tardis/plasma/equilibrium/rate_matrix.py:27]: rename

File: tardis/plasma/equilibrium/tests/test_collisional_transitions.py [hunk 4]
```diff
@@ -141,18 +141,22 @@ def test_thermal_collision_rates(
...
     )
-    coll_rates_coeff = therm_coll_rate_solver.solve([10000, 20000] * u.K)
+    coll_rates_coeff = therm_coll_rate_solver.solve([10000, 20000] * u.K).loc[
+        :, :, 0, 0, :, :
+    ]
     pdt.assert_frame_equal(
         coll_rates_coeff.iloc[:435],
         legacy_cmfgen_collision_rate_plasma_solver.coll_exc_coeff,
         check_names=False,
+        check_column_type=False,
```
Comment (wkerzendorf) [inline tardis/plasma/equilibrium/tests/test_collisional_transitions.py:151]: can you make a quick comment on that

File: tardis/plasma/equilibrium/tests/test_level_populations.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    def test_calculate_level_population_zeros(self):
+        """Test zero rate matrix."""
+        rates_matrix = np.array([[0, 0], [0, 0]])
+        with pytest.raises(np.linalg.LinAlgError):
+            self.solver._LevelPopulationSolver__calculate_level_population(
+                rates_matrix
+            )
+
+    @pytest.mark.skip(reason="Need regression data")
```
Comment (wkerzendorf) [inline tardis/plasma/equilibrium/tests/test_level_populations.py:45]: add regression data

File: tardis/plasma/equilibrium/tests/testing.ipynb
Comment (wkerzendorf) [inline tardis/plasma/equilibrium/tests/testing.ipynb:1]: delete?

Comment (wkerzendorf) [reply tardis/plasma/equilibrium/tests/testing.ipynb:1]: clean up and add it?

**Comment by andrewfullard** [comment]
Note to self: make sure the tests pull the extra atom data file from the regression data repo

File: docs/physics/plasma/equilibrium/collisional_rate_coefficients.hdf
Comment (wkerzendorf) [inline docs/physics/plasma/equilibrium/collisional_rate_coefficients.hdf:1]: I think this should be `.h5`

**Pull request merged**